src/hotspot/os/linux/memMapPrinter_linux.cpp

```diff
@@ -25,60 +25,165 @@
+      st->print("-");
+    }
+    INDENT_BY(142);
+    st->print_raw(info.filename[0] == '\0' ? "-" : info.filename);
+  #undef INDENT_BY
+    st->cr();
+  }
 
-  void print_OS_specific_details(outputStream* st) const override {
-    st->print("%s %s ", _info.prot, _info.offset);
+  void print_legend() const {
+    outputStream* st = _session.out();
+    st->print_cr("from, to, vsize: address range and size");
+    st->print_cr("prot:            protection");
+    st->print_cr("rss:             resident set size");
+    st->print_cr("hugetlb:         size of private hugetlb pages");
+    st->print_cr("pgsz:            page size");
+    st->print_cr("notes:           mapping information  (detail mode only)");
+    st->print_cr("                      shrd: mapping is shared");
+    st->print_cr("                       com: mapping committed (swap space reserved)");
+    st->print_cr("                      swap: mapping partly or completely swapped out");
+    st->print_cr("                       thp: mapping uses THP");
+    st->print_cr("                     thpad: mapping is THP-madvised");
+    st->print_cr("                     nothp: mapping is forbidden to use THP");
+    st->print_cr("                      huge: mapping uses hugetlb pages");
+    st->print_cr("vm info:         VM information (requires NMT)");
+    {
+      streamIndentor si(st, 16);
+      _session.print_nmt_flag_legend();
+    }
+    st->print_cr("file:            file mapped, if mapping is not anonymous");
   }
 
-  const char* filename() const override { return _info.filename; }
+  void print_header() const {
+    outputStream* st = _session.out();
+    //            0         1         2         3         4         5         6         7         8         9         0         1         2         3         4         5         6         7
+    //            012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789
+    //            0x0000000414000000-0x0000000453000000 123456789012 rw-p 123456789012 123456789012 16g  thp,thpadv       STACK-340754-Monitor-Deflation-Thread /shared/tmp.txt
+    st->print_cr("from               to                        vsize prot          rss      hugetlb pgsz notes            info                                  file");
+    st->print_cr("========================================================================================================================================================================");
+  }
 };
 
-void MemMapPrinter::pd_print_header(outputStream* st) {
-  st->print_cr("size          prot offset  What");
-}
-
-void MemMapPrinter::pd_iterate_all_mappings(MappingPrintClosure& closure) {
-  FILE* f = os::fopen("/proc/self/maps", "r");
+void MemMapPrinter::pd_print_all_mappings(const MappingPrintSession& session) {
+  constexpr char filename[] = "/proc/self/smaps";
```

Comment:
Is this non-constexpr if it's a `const char *` instead of `char[]`?

Reply:
not sure, tbh. sizeof is different, then, since sizeof (constexpr char[]) will give you the string size incl 0, sizeof(constexpr char*) gives you pointer size. But the pointer points to runtime addresses, so how could it be constexpr?

Reply:
>so how could it be constexpr?

Is this a nerd snipe :)? I'm fine with this being kept as is.

Reply:
> > so how could it be constexpr?
> 
> Is this a nerd snipe :)? I'm fine with this being kept as is.

? No :)

What I meant is: If I have `constexpr const char* f = "ccc"` and sizeof(f) is pointer size, I imagine there to be a pointer somewhere that points to a constant string literal in the constant segment. But if it is a real pointer, it would have to be set after program load, since only then you know the address of the string literal.

Idk. Did not investigate deeply. I like the ability of build time stringlen via sizeof for constexpr char[], that's why I prefer it.